File: formcraft-plm/src/main/java/fr/formcraft/model/entity/NonConformance.java
package fr.formcraft.model.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import fr.formcraft.model.enums.NcSeverity;
import fr.formcraft.model.enums.NcStatus;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/** A quality non-conformance raised against a product (optionally linked to a failed quality check). */
@Entity
@Table(name = "non_conformances")
@Getter
@Setter
@NoArgsConstructor
public class NonConformance {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id", nullable = false)
    @JsonIgnoreProperties({"compositionLines", "nutrientValues", "formulationResults", "workflowTasks", "qualityChecks"})
    private Product product;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "quality_check_id")
    @JsonIgnore
    private QualityCheck qualityCheck;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private NcSeverity severity;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private NcStatus status = NcStatus.OPEN;

    @Column(name = "raised_by", length = 100)
    private String raisedBy;

    @Column(name = "raised_at", nullable = false, updatable = false)
    private LocalDateTime raisedAt = LocalDateTime.now();

    @Column(name = "closed_at")
    private LocalDateTime closedAt;

    @OneToMany(mappedBy = "nonConformance", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<CorrectiveAction> correctiveActions = new ArrayList<>();
}

File: app/models/non_conformance.py
from __future__ import annotations

import datetime
from typing import List

from sqlalchemy import BigInteger, DateTime, ForeignKey, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class NonConformance(Base):
    __tablename__ = "non_conformances"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    product_id: Mapped[int] = mapped_column(BigInteger, ForeignKey("products.id", ondelete="CASCADE"), nullable=False)
    quality_check_id: Mapped[int | None] = mapped_column(BigInteger, ForeignKey("quality_checks.id", ondelete="SET NULL"))
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    severity: Mapped[str] = mapped_column(String(20), nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="OPEN")
    raised_by: Mapped[str | None] = mapped_column(String(100))
    raised_at: Mapped[datetime.datetime] = mapped_column(DateTime, nullable=False, default=lambda: datetime.datetime.now(datetime.UTC))
    closed_at: Mapped[datetime.datetime | None] = mapped_column(DateTime)

    product: Mapped["Product"] = relationship("Product")  # noqa: F821
    quality_check: Mapped["QualityCheck | None"] = relationship("QualityCheck")  # noqa: F821
    corrective_actions: Mapped[List["CorrectiveAction"]] = relationship(
        "CorrectiveAction", back_populates="non_conformance", cascade="all, delete-orphan",
    )
